File: accudrop/app/src/main/java/me/chrislane/accudrop/fragment/PlanFragment.java
package me.chrislane.accudrop.fragment;

import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.location.Location;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.PolylineOptions;

import java.util.List;

import me.chrislane.accudrop.MainActivity;
import me.chrislane.accudrop.PermissionManager;
import me.chrislane.accudrop.R;
import me.chrislane.accudrop.Util;
import me.chrislane.accudrop.presenter.PlanPresenter;
import me.chrislane.accudrop.viewmodel.GnssViewModel;
import me.chrislane.accudrop.viewmodel.RouteViewModel;

public class PlanFragment extends Fragment implements LifecycleOwner, OnMapReadyCallback {

    private static final String TAG = PlanFragment.class.getSimpleName();
    private GoogleMap map;
    private GnssViewModel gnssViewModel;
    private CameraPosition.Builder camPosBuilder;
    private RouteViewModel routeViewModel;
    private PlanPresenter planPresenter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);

        // Add this as a location listener
        if (savedInstanceState == null) {
            MainActivity main = (MainActivity) requireActivity();
            gnssViewModel = ViewModelProviders.of(main).get(GnssViewModel.class);
            routeViewModel = ViewModelProviders.of(main).get(RouteViewModel.class);
        }

        camPosBuilder = new CameraPosition.Builder()
                .zoom(15.5f)
                .bearing(0)
                .tilt(0);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_plan, container, false);

        // Set up the map
        SupportMapFragment mapFragment = (SupportMapFragment) getChildFragmentManager().findFragmentById(R.id.plan_map_fragment);
        mapFragment.getMapAsync(this);

        return view;
    }

    /**
     * Manipulates the map once available.
     * This callback is triggered when the map is ready to be used.
     * If Google Play services is not installed on the device, the user will be prompted to install
     * it inside the SupportMapFragment. This method will only be triggered once the user has
     * installed Google Play services and returned to the app.
     */
    @Override
    public void onMapReady(GoogleMap googleMap) {
        Log.i(TAG, "Map ready");
        map = googleMap;

        setupMap();
        subscribeToRoute();

        LatLng target = GnssViewModel.getLatLng(gnssViewModel.getLastLocation().getValue());
        if (target != null) {
            planPresenter = new PlanPresenter(this, target);
        } else {
            planPresenter = new PlanPresenter(this);
        }
    }

    /**
     * Set up the GoogleMap with initial settings and location.
     */
    private void setupMap() {
        MainActivity main = (MainActivity) requireActivity();
        PermissionManager permissionManager = main.getPermissionManager();

        // Initial map setup
        if (permissionManager.checkLocationPermission()) {
            map.setMyLocationEnabled(true);
        } else {
            permissionManager.requestLocationPermission("Location access is required to find your location.");
        }

        Location loc = gnssViewModel.getLastLocation().getValue();
        if (loc != null) {
            CameraPosition camPos = camPosBuilder.target(GnssViewModel.getLatLng(loc)).build();
            map.moveCamera(CameraUpdateFactory.newCameraPosition(camPos));
        }
        map.getUiSettings().setMapToolbarEnabled(false);
        map.setBuildingsEnabled(true);
        map.setMapType(GoogleMap.MAP_TYPE_HYBRID);
        map.setOnMapLongClickListener(this::onMapLongClick);
    }

    /**
     * <p>Called when the user touches and holds on the map.</p>
     * <p>This will set a new landing target for route calculations.</p>
     *
     * @param latLng The new landing target location.
     */
    private void onMapLongClick(LatLng latLng) {
        // Update map camera position
        CameraPosition camPos = camPosBuilder.target(latLng).build();
        map.animateCamera(CameraUpdateFactory.newCameraPosition(camPos));

        // Update route
        planPresenter.calcRoute(latLng);
    }

    /**
     * Subscribes to any changes to the route in the <code>RouteViewModel</code> and
     * updates the map to display these changes.
     */
    private void subscribeToRoute() {
        final Observer<List<Location>> routeObserver = route -> {
            if (route != null) {
                map.clear();

                SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getContext());
                String unitString = sharedPref.getString("general_unit", "");

                Util.Unit unit = Util.getUnit(unitString);
                for (int i = 0; i < route.size() - 1; i++) {
                    Location point1 = route.get(i);
                    Location point2 = route.get(i + 1);
                    map.addPolyline(new PolylineOptions()
                            .add(GnssViewModel.getLatLng(point1), GnssViewModel.getLatLng(point2))
                            .width(5)
                            .color(Color.RED));
                    map.addMarker(new MarkerOptions()
                            .position(GnssViewModel.getLatLng(point1))
                            .title(Util.getAltitudeText(
                                    Util.getAltitudeInUnit(point1.getAltitude(), unit), unit))
                    );
                }

                map.addMarker(new MarkerOptions()
                        .position(GnssViewModel.getLatLng(route.get(route.size() - 1)))
                        .title("Landing"));
            }
        };

        routeViewModel.getRoute().observe(this, routeObserver);
    }

    /**
     * Sets the visibility of the progress bar shown while routes are being calculated.
     *
     * @param showBar Set to <code>true</code> to show the progress bar and
     *                <code>false</code> to hide it.
     */
    public void setProgressBarVisibility(boolean showBar) {
        View view = getView();
        if (view != null) {
            ProgressBar progressBar = view.findViewById(R.id.progressbar);
            if (showBar) {
                progressBar.setVisibility(View.VISIBLE);
            } else {
                progressBar.setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/me/chrislane/accudrop/fragment/PlanFragment.kt
package me.chrislane.accudrop.fragment

import android.content.SharedPreferences
import android.graphics.Color
import android.location.Location
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import androidx.appcompat.content.res.AppCompatResources
import androidx.fragment.app.Fragment
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.PolylineOptions
import com.mapbox.maps.*
import com.mapbox.maps.plugin.LocationPuck2D
import com.mapbox.maps.plugin.locationcomponent.location
import me.chrislane.accudrop.MainActivity
import me.chrislane.accudrop.R
import me.chrislane.accudrop.UnitType
import me.chrislane.accudrop.util.DistanceAndSpeedUtil
import me.chrislane.accudrop.presenter.PlanPresenter
import me.chrislane.accudrop.viewmodel.GnssViewModel
import me.chrislane.accudrop.viewmodel.RouteViewModel
import com.mapbox.maps.CameraOptions
import com.mapbox.maps.MapView
import com.mapbox.maps.Style
import com.mapbox.maps.extension.style.expressions.dsl.generated.interpolate

class PlanFragment : Fragment(), LifecycleOwner, SharedPreferences.OnSharedPreferenceChangeListener {
    private lateinit var mapView: MapView
    private lateinit var map: MapboxMap
    private lateinit var gnssViewModel: GnssViewModel
    private lateinit var routeViewModel: RouteViewModel
    private lateinit var planPresenter: PlanPresenter
    private lateinit var preferences: SharedPreferences
    private val camPosBuilder: CameraPosition.Builder = CameraPosition.Builder().zoom(15.5f).bearing(0f).tilt(0f)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        retainInstance = true

        // Add this as a location listener
        if (savedInstanceState == null) {
            val main = requireActivity() as MainActivity
            gnssViewModel = ViewModelProvider(main).get(GnssViewModel::class.java)
            gnssViewModel.gnssListener.startListening()
            routeViewModel = ViewModelProvider(main).get(RouteViewModel::class.java)
        }

        preferences = PreferenceManager.getDefaultSharedPreferences(context)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mapView = MapView(
            inflater.context,
            MapInitOptions(inflater.context)
        )

        onMapReady(mapView)

        // Inflate the layout for this fragment
/*        val view = inflater.inflate(R.layout.fragment_plan, container, false)

        // Set up the map
        val mapFragment = childFragmentManager.findFragmentById(R.id.plan_map_fragment) as SupportMapFragment
        mapFragment.getMapAsync(this)

        preferences.registerOnSharedPreferenceChangeListener(this)*/

        return mapView
    }

    /**
     * Manipulates the map once available.
     * This callback is triggered when the map is ready to be used.
     * If Google Play services is not installed on the device, the user will be prompted to install
     * it inside the SupportMapFragment. This method will only be triggered once the user has
     * installed Google Play services and returned to the app.
     */
    fun onMapReady(mapView: MapView) {
        Log.i(TAG, "Map ready")
        map = mapView.getMapboxMap()

        setupMap()
        subscribeToRoute()

        val lastLocation = gnssViewModel.getLastLocation().value
        planPresenter = if (lastLocation != null) {
            val gnssLocation = GnssViewModel.getLatLng(lastLocation)
            PlanPresenter(this, gnssLocation)
        } else {
            PlanPresenter(this)
        }
    }

    /**
     * Set up the GoogleMap with initial settings and location.
     */
    private fun setupMap() {
        val main = requireActivity() as MainActivity
        val permissionManager = main.permissionManager

        // Initial map setup
        if (permissionManager.checkLocationPermission()) {
        } else {
            permissionManager.requestLocationPermission("Location access is required to find your location.")
            return
        }
        mapView.getMapboxMap().setCamera(
            CameraOptions.Builder()
                .zoom(14.0)
                .build()
        )
        mapView.getMapboxMap().loadStyleUri(
            Style.SATELLITE_STREETS
        ) {
            initLocationComponent()
        }
    }

    private fun initLocationComponent() {
        val main = requireActivity() as MainActivity
        val locationComponentPlugin = mapView.location
        locationComponentPlugin.updateSettings {
            this.enabled = true
            this.locationPuck = LocationPuck2D(
                bearingImage = AppCompatResources.getDrawable(
                    main,
                    R.drawable.mapbox_user_puck_icon,
                ),
                shadowImage = AppCompatResources.getDrawable(
                    main,
                    R.drawable.mapbox_user_icon_shadow,
                ),
                scaleExpression = interpolate {
                    linear()
                    zoom()
                    stop {
                        literal(0.0)
                        literal(0.6)
                    }
                    stop {
                        literal(20.0)
                        literal(1.0)
                    }
                }.toJson()
            )
        }
    }

    /**
     *
     * Called when the user touches and holds on the map.
     *
     * This will set a new landing target for route calculations.
     *
     * @param latLng The new landing target location.
     */
    private fun onMapLongClick(latLng: LatLng) {
        // Update map camera position
        val camPos = camPosBuilder.target(latLng).build()
        //map.animateCamera(CameraUpdateFactory.newCameraPosition(camPos))

        // Update route
        routeViewModel.setTarget(latLng)
        planPresenter.calcRoute(latLng)
    }

    /**
     * Subscribes to any changes to the route in the `RouteViewModel` and
     * updates the map to display these changes.
     */
    private fun subscribeToRoute() {
        routeViewModel.getRoute().observe(viewLifecycleOwner, Observer<MutableList<Location>> {
            if (it != null) {
                this.updateRoute(it)
            }
        })
    }

    override fun onDetach() {
        super.onDetach()
        preferences.registerOnSharedPreferenceChangeListener(this)
    }

    private fun updateRoute(route: MutableList<Location>) {
/*        map.clear()

        val unitString = preferences.getString("general_unit", "")!!
        val unit = UnitType.valueOf(unitString.toUpperCase())

        for (i in 0 until route.size - 1) {
            val point1 = route[i]
            val point2 = route[i + 1]
            map.addPolyline(PolylineOptions()
                    .add(GnssViewModel.getLatLng(point1), GnssViewModel.getLatLng(point2))
                    .width(5f)
                    .color(Color.RED))
            map.addMarker(MarkerOptions()
                    .position(GnssViewModel.getLatLng(point1))
                    .title(
                        DistanceAndSpeedUtil.getAltitudeText(
                            DistanceAndSpeedUtil.getAltitudeInUnit(point1.altitude, unit), unit))
            )
        }

        map.addMarker(MarkerOptions()
                .position(GnssViewModel.getLatLng(route[route.size - 1]))
                .title("Landing"))*/
    }

    /**
     * Sets the visibility of the progress bar shown while routes are being calculated.
     *
     * @param showBar Set to `true` to show the progress bar and
     * `false` to hide it.
     */
    fun setProgressBarVisibility(showBar: Boolean) {
        val view = view
        if (view != null) {
            val progressBar = view.findViewById<ProgressBar>(R.id.progressbar)
            progressBar.visibility = if (showBar) {
                View.VISIBLE
            } else {
                View.GONE
            }
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        when (key) {
            "general_unit" -> {
                val route = routeViewModel.getRoute().value
                if (route != null) {
                    updateRoute(route)
                }
            }
            "landing_pattern_downwind_altitude", "landing_pattern_crosswind_altitude", "landing_pattern_upwind_altitude" -> {
                val target = routeViewModel.getTarget().value
                if (target != null) {
                    planPresenter.calcRoute(target)
                }
            }
        }
    }

    companion object {
        private val TAG: String = PlanFragment::class.java.simpleName
    }
}
